api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/GaugeHandler.java

```diff
@@ -300,8 +307,34 @@ public void findGaugeData(
                 percentiles = new Percentiles(Collections.<Double>emptyList());
             }
 
-            metricsService.findGaugeStats(metricId, timeRange.getStart(), timeRange.getEnd(), buckets,
-                    percentiles.getPercentiles())
+            Observable<TimeRange> observableTimeRange = Observable.just(timeRange);
+
+            if (Boolean.TRUE.equals(fromEarliest)) {
+                metricsService.findMetric(metricId).first().map((metric) -> {
+                    long now = System.currentTimeMillis();
+                    long earliest = now - metric.getDataRetention() * 24 * 60 * 60 * 1000L;
+                    return new TimeRange(earliest, now);
+                });
+            }
+
+            final Percentiles localPercentiles = percentiles;
+            observableTimeRange
```

Comment:
Just above this line, there is a query to get the retention setting for the metric. The code in the if statement creates a TimeRange observable that is then fed into metric query that you highlighted. 

Also, the sample below is not good because it filters out all the empty buckets. We want to filter out only buckets that are empty at the start of the sequence; empty buckets in the middle or the end of the sequence do not need to be purged.

```
  .filter(bucket -> !bucket.isEmpty())
```